# 02 데이터셋 로드 및 전처리

**목적:** Kaggle Ecuador Favorita 원본 CSV 7개를 하나의 분석용 테이블로 통합하고, 이후 장(03~11)에서 쓸 **일별·주간** parquet를 생성합니다.

| 단계 | 산출물 | 용도 |
|------|--------|------|
| 일별 통합 | `df.parquet` (~300만 행) | 특징 엔지니어링, 일별 예측 |
| 주간 집계 | `df_weekly.parquet` (165 시계열) | SBC·클러스터링·예측 (type×family) |

**분석 단위:** store 54개가 아니라 **store type 5종(D,B,C,E,A) × family 33종 = 165 시계열**

### ⓪ 환경 설정

프로젝트 경로(`utils.paths`)를 불러오고, 원본(`data/raw`)·가공(`data/processed`) 디렉터리를 확인합니다.

In [1]:
import sys
from pathlib import Path
import pandas as pd
import numpy as np

# --- 경로 설정: 노트북이 code/ 또는 repo root에서 실행돼도 동작 ---
NOTEBOOK_DIR = Path.cwd()
if NOTEBOOK_DIR.name == 'code':
    REPO_ROOT = NOTEBOOK_DIR.parent
else:
    REPO_ROOT = NOTEBOOK_DIR
sys.path.insert(0, str(REPO_ROOT / 'code'))

# 프로젝트 공통 경로 상수 (data/raw, data/processed 등)
from utils.paths import DATA_RAW, DATA_PROCESSED, DF_DAILY, DF_WEEKLY, ensure_dirs
from utils.weekly import to_weekly_type_family

ensure_dirs()  # 출력 폴더 없으면 생성
pd.set_option('display.max_columns', None)
print('DATA_RAW:', DATA_RAW)
print('DATA_PROCESSED:', DATA_PROCESSED)


DATA_RAW: C:\Users\kjh\ai-retail-demandforecasting\data\raw
DATA_PROCESSED: C:\Users\kjh\ai-retail-demandforecasting\data\processed


### ① 판매(train) 데이터 로드

핵심 테이블 `train.csv`: 일별 × 매장 × 상품군(family) 판매량(`sales`)과 프로모션(`onpromotion`).

In [2]:
# train.csv: date × store_nbr × family 단위 일별 판매
train = pd.read_csv(DATA_RAW / 'train.csv', parse_dates=['date'])
train['date'] = pd.to_datetime(train['date'])
print('train shape:', train.shape)
train.head()


train shape: (3000888, 6)


,id,date,store_nbr,family,sales,onpromotion
0,0,2013-01-01,1,AUTOMOTIVE,0.0,0
1,1,2013-01-01,1,BABY CARE,0.0,0
2,2,2013-01-01,1,BEAUTY,0.0,0
3,3,2013-01-01,1,BEVERAGES,0.0,0
4,4,2013-01-01,1,BOOKS,0.0,0


### ② 유가(oil) 시계열 보간

`oil.csv`는 주말·휴일에 값이 비어 있습니다. 일별 빈도로 리샘플한 뒤 **선형 보간**으로 결측을 채웁니다.

In [3]:
# WTI 유가: 거래 없는 날은 NaN → 일별 리샘플 후 선형 보간
oil = pd.read_csv(DATA_RAW / 'oil.csv', parse_dates=['date'])
oil['date'] = pd.to_datetime(oil['date'])
oil = oil.sort_values('date').set_index('date').resample('D').asfreq().reset_index()
oil['dcoilwtico'] = oil['dcoilwtico'].interpolate(method='linear')
oil.head()


,date,dcoilwtico
0,2013-01-01,NaN
1,2013-01-02,93.140000
2,2013-01-03,92.970000
3,2013-01-04,93.120000
4,2013-01-05,93.146667


### ③ 공휴일·이벤트 전처리

`holidays_events.csv`를 **일별 1행**으로 집계합니다. 같은 날 여러 이벤트가 있으면 `event_types`, `descriptions`를 문자열로 합칩니다.

In [4]:
# 공휴일·이벤트: 하루에 여러 행 → date당 1행으로 집계
holidays = pd.read_csv(DATA_RAW / 'holidays_events.csv', parse_dates=['date'])
holidays['date'] = pd.to_datetime(holidays['date'])
holidays_agg = (
    holidays.groupby('date')
    .agg(
        is_holiday=('type', lambda x: (x == 'Holiday').any()),
        is_transfer=('transferred', 'any'),
        event_types=('type', lambda x: '|'.join(sorted(x.unique().astype(str)))),
        descriptions=('description', lambda x: '|'.join(x.astype(str).unique()[:3])),
    )
    .reset_index()
)
holidays_agg.head()


,date,is_holiday,is_transfer,event_types,descriptions
0,2012-03-02,True,False,Holiday,Fundacion de Manta
1,2012-04-01,True,False,Holiday,Provincializacion de Cotopaxi
2,2012-04-12,True,False,Holiday,Fundacion de Cuenca
3,2012-04-14,True,False,Holiday,Cantonizacion de Libertad
4,2012-04-21,True,False,Holiday,Cantonizacion de Riobamba


### ④ 매장·거래 데이터 로드

- `stores.csv`: 매장별 도시, **type**(D/B/C/E/A), cluster
- `transactions.csv`: 매장·일별 총 거래 건수 (수요 강도 보조 변수)

In [5]:
# 매장 메타(type=센터 유형) + 일별 거래건수
stores = pd.read_csv(DATA_RAW / 'stores.csv')
transactions = pd.read_csv(DATA_RAW / 'transactions.csv', parse_dates=['date'])
transactions['date'] = pd.to_datetime(transactions['date'])
print('stores:', stores.shape, '| transactions:', transactions.shape)
stores.head()


stores: (54, 5) | transactions: (83488, 3)


,store_nbr,city,state,type,cluster
0,1,Quito,Pichincha,D,13
1,2,Quito,Pichincha,D,13
2,3,Quito,Pichincha,D,8
3,4,Quito,Pichincha,D,9
4,5,Santo Domingo,Santo Domingo de los Tsachilas,D,4


### ⑤ 마스터 테이블 병합 (train + 외생변수)

판매(train)에 유가·공휴일·매장정보·거래건수를 날짜/매장 기준으로 left join 합니다. `type`(매장 유형)이 이후 **센터 비교 축**이 됩니다.

In [6]:
# train 기준으로 외생변수 병합 → 분석용 마스터 df
df = train.copy()
df['year'] = df['date'].dt.year
df['month'] = df['date'].dt.month
df['weekofyear'] = df['date'].dt.isocalendar().week.astype(int)

df = df.merge(oil[['date', 'dcoilwtico']], on='date', how='left')
df = df.merge(holidays_agg, on='date', how='left')
df = df.merge(stores, on='store_nbr', how='left')
df = df.merge(transactions[['date', 'store_nbr', 'transactions']], on=['date', 'store_nbr'], how='left')

# 공휴일 없는 날은 False로 채움
df['is_holiday'] = df['is_holiday'].fillna(False).astype(bool)
df['is_transfer'] = df['is_transfer'].fillna(False).astype(bool)

# 컬럼 순서 정리 (이후 장에서 쓰는 핵심 변수만)
cols = [
    'id', 'date', 'year', 'month', 'weekofyear', 'store_nbr', 'city', 'state', 'type', 'cluster',
    'family', 'sales', 'onpromotion', 'transactions', 'dcoilwtico', 'is_holiday', 'is_transfer',
    'event_types', 'descriptions'
]
df = df[[c for c in cols if c in df.columns]]
df = df.sort_values(['year', 'date', 'store_nbr', 'family']).reset_index(drop=True)
print('df shape:', df.shape)
df.head()


C:\Users\kjh\AppData\Local\Temp\ipykernel_22108\1007446337.py:13: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['is_holiday'] = df['is_holiday'].fillna(False).astype(bool)
C:\Users\kjh\AppData\Local\Temp\ipykernel_22108\1007446337.py:14: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['is_transfer'] = df['is_transfer'].fillna(False).astype(bool)


df shape: (3000888, 19)


,id,date,year,month,weekofyear,store_nbr,city,state,type,cluster,family,sales,onpromotion,transactions,dcoilwtico,is_holiday,is_transfer,event_types,descriptions
0,0,2013-01-01,2013,1,1,1,Quito,Pichincha,D,13,AUTOMOTIVE,0.0,0,NaN,NaN,True,False,Holiday,Primer dia del ano
1,1,2013-01-01,2013,1,1,1,Quito,Pichincha,D,13,BABY CARE,0.0,0,NaN,NaN,True,False,Holiday,Primer dia del ano
2,2,2013-01-01,2013,1,1,1,Quito,Pichincha,D,13,BEAUTY,0.0,0,NaN,NaN,True,False,Holiday,Primer dia del ano
3,3,2013-01-01,2013,1,1,1,Quito,Pichincha,D,13,BEVERAGES,0.0,0,NaN,NaN,True,False,Holiday,Primer dia del ano
4,4,2013-01-01,2013,1,1,1,Quito,Pichincha,D,13,BOOKS,0.0,0,NaN,NaN,True,False,Holiday,Primer dia del ano


### ⑥ 데이터 커버리지 검증

`(date, store_nbr, family)` 조합이 **누락 없이 완전 그리드**인지 확인합니다. 행 수 = 고유 조합 수이면 결측 시계열이 없다는 뜻입니다.

In [7]:
# 완전 그리드 여부: 모든 (날짜, 매장, family) 조합이 존재하는지
n_dates = df['date'].nunique()
n_combo = df.groupby(['date', 'store_nbr', 'family']).ngroups
print('날짜 수:', n_dates)
print('고유 조합 수:', n_combo, '| 총 행 수:', len(df))
print('완전 그리드:', n_combo == len(df))


날짜 수: 1684
고유 조합 수: 3000888 | 총 행 수: 3000888
완전 그리드: True


### ⑦ 일별 통합 테이블 저장

전처리 완료된 일별 df를 `data/processed/df.parquet`에 저장합니다. 03장 특징 엔지니어링의 입력 파일입니다.

In [8]:
# 일별 마스터 테이블 parquet 저장
df.to_parquet(DF_DAILY, index=False)
print('저장:', DF_DAILY)


저장: C:\Users\kjh\ai-retail-demandforecasting\data\processed\df.parquet


### ⑧ 주간 집계 (type × family)

일별 데이터를 **ISO 주차** 기준으로 합산합니다. 이후 SBC 분류·클러스터링·예측은 모두 이 **165개 주간 시계열**을 대상으로 합니다.

In [9]:
# type×family 단위 주간 합산 → 165개 시계열 생성
df_weekly = to_weekly_type_family(df)
df_weekly.to_parquet(DF_WEEKLY, index=False)
print('저장:', DF_WEEKLY)
print('주간 시계열 수 (type x family):', df_weekly.groupby(['type', 'family']).ngroups)
df_weekly.head()


저장: C:\Users\kjh\ai-retail-demandforecasting\data\processed\df_weekly.parquet
주간 시계열 수 (type x family): 165


,type,family,year,week,yearweek,sales,onpromotion,transactions,dcoilwtico,is_holiday
0,A,AUTOMOTIVE,2013,1,201301,430.0,0,139419.0,93.110000,True
1,A,AUTOMOTIVE,2013,2,201302,582.0,0,162919.0,93.538571,False
2,A,AUTOMOTIVE,2013,3,201303,582.0,0,161440.0,94.927143,False
3,A,AUTOMOTIVE,2013,4,201304,621.0,0,157677.0,95.531429,False
4,A,AUTOMOTIVE,2013,5,201305,663.0,0,165483.0,97.190000,False


## 분석 요약

| 항목 | 결과 |
|------|------|
| 일별 행 수 | 3,000,888 (date×store×family 완전 그리드) |
| 기간 | 2013-01-01 ~ 2017-08-15 (1,684일) |
| 매장 type | D, B, C, E, A (5종) |
| 주간 시계열 | **165개** (type×family) |
| 산출물 | `df.parquet`, `df_weekly.parquet` |

**다음 장(03):** 주간 시계열에 ADI, CV², lag 등 특징을 추가합니다.